File: BOJ/[BOJ 14940] 쉬운 최단거리.py
import sys, collections;
input = sys.stdin.readline

def bfs(y, x):
  q = collections.deque()
  q.append([y, x, 0]) # [y, x, distance]
  d = [[-1,0],[1,0],[0,-1],[0,1]]
  answer = [[0]*M for _ in range(N)]
  visited = [[0]*M for _ in range(N)]
  visited[y][x] = 1

  while q:
    y, x, distance = q.popleft()
    answer[y][x]=distance
    for i in range(4):
      ny, nx = y+d[i][0], x+d[i][1]
      if ny>N-1 or nx>M-1 or ny<0 or nx<0: continue
      if lst[ny][nx]==0 or visited[ny][nx]==1: continue
      visited[ny][nx]=1
      q.append([ny, nx, distance+1])
  return answer

def find_destination():
  for i in range(N):
    for j in range(M):
      if lst[i][j]==2:
        return [i, j]
  return [-1, -1]


N, M = map(int, input().split())
lst = [list(map(int, input().split())) for _ in range(N)]
y, x = find_destination()

answer = bfs(y, x)
for i in range(N):
  for j in range(M):
    if answer[i][j]==0 and lst[i][j]==1:
      answer[i][j]=-1

for i in range(N):
  print(*answer[i])

File: BOJ/[BOJ 14940] 쉬운 최단거리.js
const fs = require("fs");
const input = fs.readFileSync("/dev/stdin").toString().trim().split("\n");

const [N, M] = input.splice(0, 1)[0].split(" ").map(Number);
const arr = input.map((item) => item.split(" ").map(Number));

const bfs = (y, x) => {
  const q = [[y, x, 0]];
  const result = Array.from({ length: N }, () =>
    Array.from({ length: M }).fill(0)
  );
  const visited = Array.from({ length: N }, () =>
    Array.from({ length: M }).fill(0)
  );
  visited[y][x] = 1;

  const direct = [
    [0, 1],
    [0, -1],
    [1, 0],
    [-1, 0],
  ];

  while (q.length > 0) {
    const [y, x, distance] = q.shift();
    result[y][x] = distance;
    for (let i = 0; i < 4; i++) {
      const ny = y + direct[i][0];
      const nx = x + direct[i][1];

      if (ny > N - 1 || nx > M - 1 || ny < 0 || nx < 0) continue;
      if (arr[ny][nx] === 0) continue;
      if (visited[ny][nx]) continue;
      visited[ny][nx] = 1;
      q.push([ny, nx, distance + 1]);
    }
  }

  return result;
};

const getStartLocation = () => {
  for (let i = 0; i < N; i++) {
    for (let j = 0; j < M; j++) {
      if (arr[i][j] === 2) return [i, j];
    }
  }
  return [0, 0];
};

const [startY, startX] = getStartLocation();
const result = bfs(startY, startX);

for (let i = 0; i < N; i++) {
  for (let j = 0; j < M; j++) {
    if (arr[i][j] === 1 && result[i][j] === 0) result[i][j] = -1;
  }
}

console.log(result.map((item) => item.join(" ")).join("\n"));
